src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplate.java

```diff
@@ -579,44 +592,74 @@ private void deleteDocument(@NonNull Document document, @NonNull List<String> pa
                     final DocumentDbPageRequest pageRequest = DocumentDbPageRequest.of(pageable.getPageNumber(),
                             pageable.getPageSize(), r.getResponseContinuation());
 
-                    return new PageImpl<>(result, pageRequest, count(query, domainClass, collectionName));
+                    return new PageImpl<>(result, pageRequest, count);
                 });
     }
 
+    // Internal class to wrap count and FeedResponse data
+    @AllArgsConstructor
+    private static class PageResponse<T extends Resource> {
+        @Getter @Setter
+        private long count;
+        @Getter @Setter
+        private FeedResponse<T> response;
+    }
+
     @Override
     public long count(String collectionName) {
         Assert.hasText(collectionName, "collectionName should not be empty");
 
         final DocumentQuery query = new DocumentQuery(Criteria.getInstance(CriteriaType.ALL));
-        final com.microsoft.azure.documentdb.SqlQuerySpec querySpec = new CountQueryGenerator().generate(query);
-
-        return getCountValue(querySpec, true, collectionName);
+        return toCountValue(getCountValue(query, true, collectionName));
     }
 
     @Override
     public <T> long count(DocumentQuery query, Class<T> domainClass, String collectionName) {
-        Assert.notNull(domainClass, "domainClass should not be null");
-        Assert.hasText(collectionName, "collectionName should not be empty");
-
-        final com.microsoft.azure.documentdb.SqlQuerySpec querySpec = new CountQueryGenerator().generate(query);
         final boolean isCrossPartitionQuery = query.isCrossPartitionQuery(getPartitionKeyNames(domainClass));
-
-        return getCountValue(querySpec, isCrossPartitionQuery, collectionName);
+        return toCountValue(getCountValue(query, isCrossPartitionQuery, collectionName));
     }
 
-    private long getCountValue(com.microsoft.azure.documentdb.SqlQuerySpec querySpec, boolean isCrossPartitionQuery,
-                               String collectionName) {
-        final com.microsoft.azure.documentdb.FeedResponse<Document> feedResponse =
-                executeQuery(querySpec, isCrossPartitionQuery, collectionName);
-        final Object value = feedResponse.getQueryIterable().toList().get(0).getHashMap().get(COUNT_VALUE_KEY);
-
-        if (value instanceof Integer) {
-            return Long.valueOf((Integer) value);
-        } else if (value instanceof Long) {
-            return (Long) value;
-        } else {
-            throw new IllegalStateException("Unexpected value type " + value.getClass() + " of value: " + value);
+    private Long toCountValue(Observable<Long> observable) {
+        final List<Long> result = Lists.newArrayList();
```

Comment:
1. I'm confusing here why use count down latch. If you want to blocking get the vaule, please try `observable.toBlocking().first()`.
2. I'm unsure whether this might be accessed by multiple threads, if so, result of type `List` is not thread safe.

Reply:
You can just blocking get of `getCountValue`